classes/log/Repository.php

```diff
@@ -0,0 +1,187 @@
+                );
+            }
+        }
+    }
+
+    /**
+     * Delete all log entries for an object.
+     *
+     * @param int $assocType
+     * @param int $assocId
+     * @return int The number of affected rows.
+     */
+    public function deleteByAssoc(int $assocType, int $assocId): int
+    {
+        return $this->model
+            ->newQuery()
+            ->where('assoc_type', (int)$assocType)
+            ->where('assoc_id', (int)$assocId)
+            ->delete();
+    }
+
+    /**
+     * Transfer all log and log users entries to another user.
+     *
+     * @param int $oldUserId
+     * @param int $newUserId
+     */
+    public function changeUser($oldUserId, $newUserId)
+    {
+        return [
+            DB::update(
+                'UPDATE email_log SET sender_id = ? WHERE sender_id = ?',
+                [(int) $newUserId, (int) $oldUserId]
+            ),
+            DB::update(
+                'UPDATE email_log_users
+                SET user_id = ?
+                WHERE user_id = ? AND email_log_id NOT IN (SELECT t1.email_log_id
+                    FROM (SELECT email_log_id FROM email_log_users WHERE user_id = ?) AS t1
+                    INNER JOIN
+                    (SELECT email_log_id FROM email_log_users WHERE user_id = ?) AS t2
+                    ON t1.email_log_id = t2.email_log_id);',
+                [(int)$newUserId, (int)$oldUserId, (int)$newUserId, (int)$oldUserId]
+            )
+        ];
+    }
+
+    /**
+     * Create a log entry for a submission from data in a Mailable class
+     *
+     * @param EmailLogEventType $eventType Type of email event
+     * @param Mailable $mailable
+     * @param Submission $submission
+     * @param ?User $sender
+     * @return int The new log entry id
+     */
+    function logMailable(EmailLogEventType $eventType, Mailable $mailable, Submission $submission, ?User $sender = null): int
+    {
+        $clonedMailable = clone $mailable;
+        $clonedMailable->removeFooter();
+
+        $this->model->eventType = $eventType->value;
+        $this->model->assocId = $submission->getId();
+        $this->model->dateSent = Core::getCurrentDate();
+        $this->model->senderId = $sender ? $sender->getId() : null;
+        $this->model->fromAddress = $this->getContactString($clonedMailable->from);
+        $this->model->recipients = $this->getContactString($clonedMailable->to);
+        $this->model->ccRecipients = $this->getContactString($clonedMailable->cc);
+        $this->model->bccRecipients = $this->getContactString($clonedMailable->bcc);
+        $this->model->body = $clonedMailable->render();
+        $this->model->assocType = Application::ASSOC_TYPE_SUBMISSION;
+        $this->model->subject = Mail::compileParams(
+            $clonedMailable->subject,
+            $clonedMailable->getData(Locale::getLocale())
+        );
+
+        $this->model->save();
+        $this->insertLogUserIds( $this->model);
```

Comment:
The space after the `(` suggests the automatic code formatter isn't configured. This should be running before commits are allowed. Check out the documentation for Captain Hook, which is used to kick this off in pkp-lib.